src/wms/src/main/java/org/geoserver/wms/clip/ClipWMSGetMapCallBack.java

```diff
@@ -0,0 +1,176 @@
+/* (c) 2019 Open Source Geospatial Foundation - all rights reserved
+ * This code is licensed under the GPL 2.0 license, available at the root
+ * application directory.
+ */
+package org.geoserver.wms.clip;
+
+import java.awt.Rectangle;
+import java.awt.geom.AffineTransform;
+import java.awt.geom.Point2D;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.regex.Pattern;
+import org.apache.commons.beanutils.BeanUtilsBean2;
+import org.geoserver.wms.FeatureInfoRequestParameters;
+import org.geoserver.wms.GetMapCallback;
+import org.geoserver.wms.GetMapRequest;
+import org.geoserver.wms.WMSMapContent;
+import org.geoserver.wms.WebMap;
+import org.geotools.geometry.DirectPosition2D;
+import org.geotools.geometry.jts.JTS;
+import org.geotools.geometry.jts.ReferencedEnvelope;
+import org.geotools.geometry.jts.WKTReader2;
+import org.geotools.map.FeatureLayer;
+import org.geotools.map.GridReaderLayer;
+import org.geotools.map.Layer;
+import org.geotools.referencing.CRS;
+import org.geotools.renderer.lite.RendererUtilities;
+import org.geotools.util.logging.Logging;
+import org.locationtech.jts.geom.Geometry;
+import org.locationtech.jts.geom.MultiPolygon;
+import org.locationtech.jts.geom.Polygon;
+import org.opengis.referencing.crs.CoordinateReferenceSystem;
+import org.opengis.referencing.operation.MathTransform;
+
+/** @author ImranR */
+public class ClipWMSGetMapCallBack implements GetMapCallback {
+
+    //  private static Set<ProcessFactory> processFactories = Processors.getProcessFactories();
+    private static final Logger LOGGER = Logging.getLogger(ClipWMSGetMapCallBack.class.getName());
+
+    private static final WKTReader2 reader = new WKTReader2();
+
+    private static final Pattern SRID_REGEX = Pattern.compile("SRID=[0-9].*");
+
+    @Override
+    public GetMapRequest initRequest(GetMapRequest request) {
+        return request;
+    }
+
+    @Override
+    public void initMapContent(WMSMapContent mapContent) {}
+
+    @Override
+    public Layer beforeLayer(WMSMapContent mapContent, Layer layer) {
+
+        // find and parse WKT geometry from WMS request
+        //
+        Geometry wktGeom = getClipGeometry(mapContent.getRequest());
+        if (wktGeom == null) return layer;
+
+        Geometry bboxGeom = JTS.toGeometry(mapContent.getRequest().getBbox());
+        // check: if wkt area fully contains bbox
+        if (wktGeom.covers(bboxGeom)) return layer;

```

Comment:
Does it work fine if the request is a WMS 1.3, typically with a flipped bbox?  I don't see tests covering this case (clip fully containing map area). Please choose a bbox that would not be a match for the clip if the flipped version is used. (e.g., not 20,20,40,40 :-D ).

Reply:
there are WMS 1.3.0 unit tests also with flipped BBOX